File: 05월 17일 - 트리/과제/15681.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**8)

"""
 PPT 트리의 정점의 수 구하기 응용

 [트리와 쿼리]

 1. 루트에서부터 dfs 탐색
 2. 각 노드를 루트로 하는 서브트리의 정점 수를 재귀적으로 계산
    - 서브트리에 속한 정점의 개수를 저장하는 dp 배열의 초기화를 -1로 해주고, dfs 탐색 시 현재 정점의 dp 값을 0으로 설정함으로써 자식 노드만 탐색할 수 있도록 함 (부모 노드에 0이 저장되어 있으므로 바로 리턴)
"""

def tree_dp(curr, graph):
    if subtree_cnt[curr] != -1:
        return subtree_cnt[curr]

    # 서브트리에 속한 정점의 수를 세고, 저장
    subtree_cnt[curr] = 0   # 현재 노드를 0으로 표기하여 부모 자식간 계속 호출되지 않도록
    cnt = 1
    for next in graph[curr]:
        cnt += tree_dp(next, graph)

    subtree_cnt[curr] = cnt
    return cnt


# 입력
n, r, q = map(int, input().split())
graph = [list() for _ in range(n+1)]
subtree_cnt = [-1]*(n+1)

for _ in range(n-1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

# 연산
tree_dp(r, graph)


# 출력
for _ in range(q):
    print(subtree_cnt[int(input())])

File: 05월 17일 - 트리/과제/15681.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> subtree_cnt;

int treeDp(int cur, vector<vector<int>> &graph) {
    if (subtree_cnt[cur] != -1) { //이미 탐색했던 상태
        return subtree_cnt[cur];
    }

    //서브트리에 속한 정점의 수를 세고, 저장
    subtree_cnt[cur] = 0;
    int node = 1;
    for (int i = 0; i < graph[cur].size(); i++) {
        node += treeDp(graph[cur][i], graph);
    }
    return subtree_cnt[cur] = node;
}

/**
 * PPT 트리의 정점의 수 구하기 응용
 *
 * [트리와 쿼리]
 *
 * 1. 루트에서부터 dfs 탐색
 * 2. 각 노드를 루트로 하는 서브트리의 정점 수를 재귀적으로 메모리제이션하며 계산
 * - 서브트리에 속한 정점의 개수를 저장하는 dp 배열의 초기화를 -1로 해주고,
 *   dfs 탐색 시 현재 정점의 dp 값을 0으로 설정함으로써 자식 노드만 탐색할 수 있도록 함 (부모 노드에 0이 저장되어 있으므로 바로 리턴)
 *
 */

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, r, q, u, v, input;

    //입력
    cin >> n >> r >> q;
    vector<vector<int>> graph(n + 1, vector<int>(0));
    subtree_cnt.assign(n + 1, -1);

    while (--n) { //무방향 그래프
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    //연산
    treeDp(r, graph);

    //출력
    while (q--) {
        cin >> input;
        cout << subtree_cnt[input] << '\n';
    }
    return 0;
}